crank/tests/test_stack_api.py

```diff
@@ -0,0 +1,113 @@
+from geometric.nifty import ang2bohr
+from crank import crankServerAPI
+
+try:
+    import qcengine
+    import psi4
+except:
+    pass
+
+class SimpleServer:
+    """ A simple server that interfaces with crank and geometric to do the dihedral scanning work flow """
+    def __init__(self, xyzfilename, dihedrals, grid_spacing):
+        self.M = geometric.molecule.Molecule(xyzfilename)
+        self.dihedrals = dihedrals
+        self.grid_spacing = grid_spacing
+        self.elements = self.M.elem
+        self.init_coords = [(self.M.xyzs[0] * ang2bohr).ravel().tolist()]
+
+    def run_crank_scan(self):
+        """
+        Run crank scan in the following steps:
+        1. Create json input for crank
+        2. Send the json input dictionary to crankAPI.crank_api(), get the next set of jobs
+        3. If there are no jobs needed, finish and return the lowest energy on each dihedral grid
+        4. If there are new jobs, run them with geomeTRIC.run_json.
+        5. Collect the results and put them into new json input dictionary
+        6. Go back to step 2.
+        """
+        # step 1
+        crank_state = crankServerAPI.create_initial_api_input(dihedrals=self.dihedrals, grid_spacing=self.grid_spacing, elements=self.elements, init_coords=self.init_coords)
+        while True:
+            # step 2
+            next_jobs = crank_api(crank_state, verbose=True)
+            # step 3
+            if len(next_jobs) == 0:
+                print("Crank Scan Finished")
+                return crankServerAPI.collect_lowest_energies(crank_state)
+            # step 4
+            job_results = collections.defaultdict(list)
+            for grid_id_str, job_geo_list in next_jobs.items():
+                for job_geo in job_geo_list:
+                    dihedral_values = tuple(int(i) for i in grid_id_str.split(','))
+                    geometric_input_dict = self.make_geomeTRIC_input(dihedral_values, job_geo)
+                    geometric_output_dict = geometric.run_json.geometric_run_json(geometric_input_dict)
+                    final_geo = geometric_output_dict['final_molecule']['molecule']['geometry']
+                    final_energy = geometric_output_dict['final_molecule']['properties']['return_energy']
+                    # Note: the results should be appended in the same order as in the inputs
+                    # It's not a problem here when running serial for loop
+                    job_results[grid_id_str].append((job_geo, final_geo, final_energy))
+            # step 5
+            crank_state = crankServerAPI.update_crank_state(crank_state, job_results)
+            # step 6
+
+
+    def make_geomeTRIC_input(self, dihedral_values, geometry):
+        """ This function should be implemented on the server, that takes QM specs, geometry and constraint
+        to generate a geomeTRIC json input dictionary"""
+        constraints_string = "$set\n"
+        for (d1, d2, d3, d4), v in zip(self.dihedrals, dihedral_values):
```

Comment:
The representation of dihedral angles here are sitting on the server. I think there is no way the server can blindly host a dihedral scan without even known what are the dihedral angles. The representation here doesn't have to match what crank uses, but I just used the same representation to be convenient.

Reply:
I mean my comment was similar to above that we should try to encode these as something besides a string and passed into geomeTRIC via a list of tuples for examples. This will work for now, but I would strongly suggest looking into fixing this for geomeTRIC.